src/Features/Core/Portable/CodeRefactorings/AddMissingImports/AbstractAddMissingImportsFeatureService.cs

```diff
@@ -0,0 +1,150 @@
+            if (usableFixes.IsEmpty)
+            {
+                return document.Project;
+            }
+
+            // Apply those fixes to the document.
+            var newDocument = await ApplyFixesAsync(document, usableFixes, cancellationToken).ConfigureAwait(false);
+            return newDocument.Project;
+        }
+
+        private async Task<ImmutableArray<Diagnostic>> GetDiagnosticsAsync(Document document, TextSpan textSpan, CancellationToken cancellationToken)
+        {
+            var project = document.Project;
+            var compilation = await project.GetCompilationAsync(cancellationToken);
+
+            // Ensure that diagnostics for this document are always in document location
+            // order.  This provides a consistent and deterministic order for fixers
+            // that want to update a document.
+            var diagnostics = compilation.GetDiagnostics(cancellationToken)
+                .Where(diagnostic => FixableDiagnosticIds.Contains(diagnostic.Id)
+                    && diagnostic.Location.SourceTree.FilePath == document.FilePath
+                    && diagnostic.Location.SourceSpan.IntersectsWith(textSpan))
+                .ToImmutableArray()
+                .Sort((d1, d2) => d1.Location.SourceSpan.Start - d2.Location.SourceSpan.Start);
+
+            return diagnostics;
+        }
+
+        private async Task<ImmutableArray<AddImportFixData>> GetUnambiguousFixesAsync(Document document, ImmutableArray<Diagnostic> diagnostics, CancellationToken cancellationToken)
+        {
+            var solution = document.Project.Solution;
+            var symbolSearchService = solution.Workspace.Services.GetService<ISymbolSearchService>();
+            // Since we are not currently considering NuGet packages, pass an empty array
+            var packageSources = ImmutableArray<PackageSource>.Empty;
+            var addImportService = document.GetLanguageService<IAddImportFeatureService>();
+
+            var getFixesForDiagnosticsTasks = diagnostics
+                .GroupBy(diagnostic => diagnostic.Location.SourceSpan)
+                .Select(diagnosticsForSourceSpan => addImportService
+                    .GetFixesForDiagnosticsAsync(document, diagnosticsForSourceSpan.Key, diagnosticsForSourceSpan.AsImmutable(),
+                        symbolSearchService, searchReferenceAssemblies: true, packageSources, cancellationToken));
+
+            await Task.WhenAll(getFixesForDiagnosticsTasks).ConfigureAwait(false);
+
+            var fixes = ArrayBuilder<AddImportFixData>.GetInstance();
+            foreach (var getFixesForDiagnosticsTask in getFixesForDiagnosticsTasks)
+            {
+                var fixesForDiagnostics = await getFixesForDiagnosticsTask.ConfigureAwait(false);
+
+                foreach (var fixesForDiagnostic in fixesForDiagnostics)
+                {
+                    // If there is more than one fix, then we will leave it for the user to manually apply.
```

Comment:
i'm on the fence abou tthsi comment.  i totally understand it... but i feel like it could use a bit more explanation for someone else.  it's missing the critical "why" part that helps explain why we think this is necessayr.